### PR TITLE
Introduce ChainIndepState

### LINKED ISSUE #2058: Make `nodeForgeBlock` pure
The only remaining side effect we still have is `MonadRandom`, but that is now nearly gone from consensus (#2036). Once it is gone completely, `nodeForgeBlock` can become a pure function. Once that is done, we should simplify `WithEarllyExit`; it won't need to be a monad transformer anymore.

### PR DESCRIPTION
**Pull request by mrBliss**
Closes #2058, #2229.

* Get rid of the `TPraosForgeState` wrapper around `HotKey` and track more.
  information about the `HotKey`.
* Rename `ConsensusState` to `ChainDepState`.
* Introduce `ChainIndepState` and use it to track the `HotKey` for `TPraos`.
* Rename `ForgeState` to `ExtraForgeState` (block-specific, unused, but for
  future use, e.g., hardware devices, online opcert renewal) and define
  `ForgeState` to be the product of `ChainIndepState` and `ExtraForgeState`.
* Pass the `ChainIndepState` to `checkIsLeader` so that `TPraos` can check the
  validity of the KES key (original goal of this PR).
* Remove `Update`.
* Make `forgeBlock` pure by removing `MonadRandom` (work towards #2036).
* Also remove `MonadRandom` from `ProtocolInfo`.

### CODE REVIEW

File: ouroboros-consensus/test-consensus/Test/Consensus/HardFork/Combinator.hs [hunk 1]
```diff
@@ -272,6 +267,10 @@ prop_simple_hfc_convergence testSetup@TestSetup{..} =
                 :* (WrapPartialConsensusConfig $ consensusConfigB nid)
                 :* Nil
             }
+        , configIndep  = PerEraChainIndepStateConfig $
```
Comment (edsko) [inline ouroboros-consensus/test-consensus/Test/Consensus/HardFork/Combinator.hs:270]: Use `trivial`?

File: ouroboros-consensus/test-consensus/Test/Consensus/HardFork/Combinator/A.hs [hunk 2]
```diff
@@ -102,23 +102,26 @@ data instance ConsensusConfig ProtocolA = CfgA {
...
-  type ConsensusState ProtocolA = ()
-  type LedgerView     ProtocolA = ()
-  type IsLeader       ProtocolA = ()
-  type CanBeLeader    ProtocolA = ()
-  type CannotLead     ProtocolA = Void
-  type ValidateView   ProtocolA = ()
-  type ValidationErr  ProtocolA = Void
-
-  checkIsLeader CfgA{..} () (Ticked slot _) _ =
+  type ChainDepState   ProtocolA = ()
```
Comment (edsko) [inline ouroboros-consensus/test-consensus/Test/Consensus/HardFork/Combinator/A.hs:109]: Undo realignment.

File: ouroboros-consensus/test-consensus/Test/Consensus/HardFork/Combinator/B.hs [hunk 3]
```diff
@@ -89,23 +89,26 @@ data instance ConsensusConfig ProtocolB = CfgB {
 instance ChainSelection ProtocolB where
   -- Use defaults
 
+instance HasChainIndepState ProtocolB where
+  -- Use defaults
+
 instance ConsensusProtocol ProtocolB where
-  type ConsensusState ProtocolB = ()
```
Comment (edsko) [inline ouroboros-consensus/test-consensus/Test/Consensus/HardFork/Combinator/B.hs:93]: Undo realignment.

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs [hunk 4]
```diff
@@ -419,7 +427,7 @@ protocolInfoCardano genesisByron mSigThresh pVer sVer mbCredsByron
         -- TODO 'ForgeState' is an 'NP' while 'CanBeLeader' is an 'OptNP', so
         -- even when we have no Shelley credentials, we'd still need to
         -- maintain a Shelley 'ForgeState', for which we'd need the key stored
-        -- in the credentials.
+        -- in the credentials. TODO update
```
Comment (edsko) [inline ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs:430]: Stray comment

File: ouroboros-consensus/ouroboros-consensus-mock/src/Ouroboros/Consensus/Mock/Protocol/Praos.hs [hunk 5]
```diff
@@ -250,18 +246,22 @@ data instance ConsensusConfig (Praos c) = PraosConfig
...
   protocolSecurityParam = praosSecurityParam . praosParams
 
-  type LedgerView     (Praos c) = StakeDist
-  type IsLeader       (Praos c) = PraosProof c
-  type ValidationErr  (Praos c) = PraosValidationError c
-  type ValidateView   (Praos c) = PraosValidateView    c
-  type ConsensusState (Praos c) = [BlockInfo c]
-  type CanBeLeader    (Praos c) = CoreNodeId
-  type CannotLead     (Praos c) = Void
+  type LedgerView      (Praos c) = StakeDist
```
Comment (edsko) [inline ouroboros-consensus/ouroboros-consensus-mock/src/Ouroboros/Consensus/Mock/Protocol/Praos.hs:256]: Undo realignment

File: ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Degenerate.hs [hunk 6]
```diff
@@ -185,44 +185,51 @@ data DegenForkProtocol b
...
-  -- would be /bad orphans/, i.e., for @HardForkConsensusState '[b]@.
-  type ConsensusState (DegenForkProtocol b) = DegenForkConsensusState b
-  type ValidationErr  (DegenForkProtocol b) = ValidationErr (HardForkProtocol '[b])
-  type LedgerView     (DegenForkProtocol b) = LedgerView    (HardForkProtocol '[b])
-  type CanBeLeader    (DegenForkProtocol b) = CanBeLeader   (HardForkProtocol '[b])
-  type CannotLead     (DegenForkProtocol b) = CannotLead    (HardForkProtocol '[b])
-  type IsLeader       (DegenForkProtocol b) = IsLeader      (HardForkProtocol '[b])
-  type ValidateView   (DegenForkProtocol b) = ValidateView  (HardForkProtocol '[b])
+  -- would be /bad orphans/, i.e., for @HardForkChainDepState '[b]@.
+  type ChainDepState   (DegenForkProtocol b) = DegenForkChainDepState b
```
Comment (edsko) [inline ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Degenerate.hs:216]: Unnecessary realignment

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/BlockProduction.hs [hunk 7]
```diff
@@ -55,38 +55,49 @@ blockProductionIO :: forall blk. (BlockSupportsProtocol blk, CanForge blk)
...
+  -> m (LeaderCheck     (BlockProtocol blk))
+defaultGetLeaderProof cfg proof mfs varForgeState tracer lgrSt chainDepSt = do
+    chainIndepSt <- modifyMVar varForgeState $ \forgeState -> do
+      forgeState' <-
+        updateForgeState
+          mfs
+          (configIndep cfg)
+          (tickedSlotNo lgrSt)
+          forgeState
+      traceWith tracer forgeState'
```
Comment (edsko) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node/BlockProduction.hs:96]: Let's pull out the `traceWith`.

File: ouroboros-consensus/src/Ouroboros/Consensus/Protocol/BFT.hs [hunk 8]
```diff
@@ -124,32 +124,35 @@ data instance ConsensusConfig (Bft c) = BftConfig {
...
-  type ValidateView   (Bft c) = BftValidateView c
-  type LedgerView     (Bft c) = ()
-  type IsLeader       (Bft c) = ()
-  type ConsensusState (Bft c) = ()
-  type CanBeLeader    (Bft c) = CoreNodeId
-  type CannotLead     (Bft c) = Void
+  type ValidationErr   (Bft c) = BftValidationErr
+  type ValidateView    (Bft c) = BftValidateView c
+  type LedgerView      (Bft c) = ()
+  type IsLeader        (Bft c) = ()
```
Comment (edsko) [inline ouroboros-consensus/src/Ouroboros/Consensus/Protocol/BFT.hs:137]: Alignment

File: ouroboros-consensus/src/Ouroboros/Consensus/Protocol/LeaderSchedule.hs [hunk 9]
```diff
@@ -89,26 +92,26 @@ data instance ConsensusConfig (WithLeaderSchedule p) = WLSConfig
...
 
 instance ConsensusProtocol p => ConsensusProtocol (WithLeaderSchedule p) where
-  type ConsensusState (WithLeaderSchedule p) = ()
-  type LedgerView     (WithLeaderSchedule p) = ()
-  type ValidationErr  (WithLeaderSchedule p) = ()
-  type IsLeader       (WithLeaderSchedule p) = ()
-  type ValidateView   (WithLeaderSchedule p) = ()
-  type CanBeLeader    (WithLeaderSchedule p) = ()
-  type CannotLead     (WithLeaderSchedule p) = Void
+  type ChainDepState   (WithLeaderSchedule p) = ()
```
Comment (edsko) [inline ouroboros-consensus/src/Ouroboros/Consensus/Protocol/LeaderSchedule.hs:95]: Alignment

File: ouroboros-consensus/src/Ouroboros/Consensus/Protocol/ModChainSel.hs [hunk 10]
```diff
@@ -39,24 +39,35 @@ instance ChainSelection s => ChainSelection (ModChainSel p s) where
...
-    type ConsensusState (ModChainSel p s) = ConsensusState p
-    type IsLeader       (ModChainSel p s) = IsLeader       p
-    type CanBeLeader    (ModChainSel p s) = CanBeLeader    p
-    type CannotLead     (ModChainSel p s) = CannotLead     p
-    type LedgerView     (ModChainSel p s) = LedgerView     p
-    type ValidationErr  (ModChainSel p s) = ValidationErr  p
-    type ValidateView   (ModChainSel p s) = ValidateView   p
-
-    checkIsLeader cfg canBeLeader ledgerView consensusState =
+    type ChainDepState   (ModChainSel p s) = ChainDepState   p
```
Comment (edsko) [inline ouroboros-consensus/src/Ouroboros/Consensus/Protocol/ModChainSel.hs:50]: Alignment

File: ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT.hs [hunk 11]
```diff
@@ -271,26 +271,29 @@ instance PBftCrypto c => ChainSelection (PBft c) where
...
   -- | We require two things from the ledger state:
   --
   --   - Protocol parameters, for the signature window and threshold.
   --   - The delegation map.
-  type LedgerView     (PBft c) = PBftLedgerView c
-  type IsLeader       (PBft c) = PBftIsLeader   c
-  type ConsensusState (PBft c) = PBftState      c
-  type CanBeLeader    (PBft c) = PBftIsLeader   c
-  type CannotLead     (PBft c) = PBftCannotLead c
+  type LedgerView      (PBft c) = PBftLedgerView c
```
Comment (edsko) [inline ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT.hs:285]: Alignment

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Protocol.hs [hunk 12]
```diff
@@ -230,21 +235,44 @@ data TPraosProof c = TPraosProof {
 
 instance TPraosCrypto c => NoUnexpectedThunks (TPraosProof c)
 
+data TPraosUnusableKey = TPraosUnusableKey {
+      tpraosUnusableKeyStart   :: SL.KESPeriod
```
Comment (edsko) [inline ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Protocol.hs:239]: Bangs?

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Protocol.hs [hunk 13]
```diff
@@ -230,21 +235,44 @@ data TPraosProof c = TPraosProof {
...
 -- | Expresses that, whilst we believe ourselves to be a leader for this slot,
 -- we are nonetheless unable to forge a block.
 data TPraosCannotLead c =
-    -- | Our operational certificate is not valid for the current KES period. The
-    --   given parameters are the current KES period, the minimum and maximum KES
-    --   periods for which this operational certificate is valid.
-    TPraosCannotLeadInvalidOcert KES.Period KES.Period KES.Period
+    -- | The KES key in our operational certificate is not usable for the
+    -- current KES period.
+    TPraosCannotLeadUnusableKESKey TPraosUnusableKey
```
Comment (edsko) [inline ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Protocol.hs:267]: If so, then here as well.

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Protocol.hs [hunk 14]
```diff
@@ -320,25 +360,24 @@ instance TPraosCrypto c => ConsensusProtocol (TPraos c) where
        -- The given genesis key has authority to produce a block in this
         -- slot. Check whether we're its delegate.
         Just (SL.ActiveSlot gkhash) -> case Map.lookup gkhash dlgMap of
+            Nothing
+              -> error "unknown genesis key in overlay schedule"
```
Comment (edsko) [inline ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Protocol.hs:364]: @nc6 can you confirm that this change is indeed okay?

**Comment by mrBliss** [comment]
bors merge

**Pull request merged**